ElementsProject/lightning
# Handle blockheight disagreement between us and payee

Fixes #3367 

Due to the recent BOLT change that [merges `final_expiry_too_soon` into `incorrect_or_unknown_payment_details`](https://github.com/lightningnetwork/lightning-rfc/commit/6729755f0c0228df0686176e99032fb368ca056a), the previously-retried transient failure `final_expiry_too_soon` (17), which used to be how blockheight disagreements between payee and payer were signalled, is now a permanent failure `incorrect_or_unknown_payment_details` (PERM|15).  Since it is a permanent failure at the payee end, `pay` did not retry it.  (The commit linked is factually incorrect as well: it claims `final_expiry_too_soon` to be PERM|17, but looking at its actual diff shows it was 17 before the commit).

Blockheight disagreements between payer and payee are transient and should be retried.  This patch is parses the returned `height` from the payee and checks if it is the same as the height at which we started the attempt; if different, we wait until we reach the payee blockheight using the new `waitblockheight` command (the first commit), then retry.

@NicolasDorier please notice this PR.

Currently draft since I still have to make a proper reproducible test for this.  PRed now so that @NicolasDorier can go patch it in his software.

Test plan:

* Start a `line_graph` of `l1`, `l2`, `l3`.
* Set up `l2` so it stops before it can forward any payment (I believe we have this feature in our test suite somehow, just need to review it again).
* Get an invoice from `l3`.
* In a background executor, have `l1` pay the invoice.
* Wait for `l2` to stop.
* Advance the blockchain by 1 block.
* Restart `l2`, setting it up so it will now correctly forward payments.
* Check that the background executor now completes successfully.
